File: 30_06_2024_Remove_Max_Number_of_Edges_to_Keep_Graph_Fully_Traversable.py
from typing import List

class UnionFind:
    def __init__(self, n: int):
        self.components = n
        self.parent = list(range(n + 1))
        self.size = [1] * (n + 1)

    def find(self, x: int) -> int:
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def unite(self, x: int, y: int) -> bool:
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return False
        if self.size[root_x] < self.size[root_y]:
            root_x, root_y = root_y, root_x
        self.parent[root_y] = root_x
        self.size[root_x] += self.size[root_y]
        self.components -= 1
        return True

    def is_connected(self) -> bool:
        return self.components == 1

class Solution:
    def maxNumEdgesToRemove(self, n: int, edges: List[List[int]]) -> int:
        alice, bob = UnionFind(n), UnionFind(n)
        edges_required = 0

        # Process type 3 edges first
        for edge in edges:
            if edge[0] == 3:
                if alice.unite(edge[1], edge[2]) or bob.unite(edge[1], edge[2]):
                    edges_required += 1

        # Process type 1 and type 2 edges
        for edge in edges:
            if edge[0] == 1:
                if alice.unite(edge[1], edge[2]):
                    edges_required += 1
            elif edge[0] == 2:
                if bob.unite(edge[1], edge[2]):
                    edges_required += 1

        # Check if both are fully connected
        if alice.is_connected() and bob.is_connected():
            return len(edges) - edges_required
        return -1

File: 30_06_2024_Remove_Max_Number_of_Edges_to_Keep_Graph_Fully_Traversable.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int maxNumEdgesToRemove(int n, List<List<Integer>> edges) {
        class UnionFind {
            int[] parent;
            int[] size;
            int components;

            UnionFind(int n) {
                components = n;
                parent = new int[n + 1];
                size = new int[n + 1];
                for (int i = 0; i <= n; ++i) {
                    parent[i] = i;
                    size[i] = 1;
                }
            }

            int find(int x) {
                if (parent[x] != x) {
                    parent[x] = find(parent[x]);
                }
                return parent[x];
            }

            boolean unite(int x, int y) {
                int rootX = find(x), rootY = find(y);
                if (rootX == rootY) return false;
                if (size[rootX] < size[rootY]) {
                    int temp = rootX;
                    rootX = rootY;
                    rootY = temp;
                }
                parent[rootY] = rootX;
                size[rootX] += size[rootY];
                components--;
                return true;
            }

            boolean isConnected() {
                return components == 1;
            }
        }

        UnionFind alice = new UnionFind(n);
        UnionFind bob = new UnionFind(n);
        int edgesRequired = 0;

        // Process type 3 edges first
        for (List<Integer> edge : edges) {
            if (edge.get(0) == 3) {
                if (alice.unite(edge.get(1), edge.get(2)) | bob.unite(edge.get(1), edge.get(2))) {
                    edgesRequired++;
                }
            }
        }

        // Process type 1 and type 2 edges
        for (List<Integer> edge : edges) {
            if (edge.get(0) == 1) {
                if (alice.unite(edge.get(1), edge.get(2))) {
                    edgesRequired++;
                }
            } else if (edge.get(0) == 2) {
                if (bob.unite(edge.get(1), edge.get(2))) {
                    edgesRequired++;
                }
            }
        }

        // Check if both are fully connected
        if (alice.isConnected() && bob.isConnected()) {
            return edges.size() - edgesRequired;
        }

        return -1;
    }
}

